### PR TITLE
bump to commons 3.0.6; add LocalStorageBackedAsyncStore (FF-1980)

### PR DESCRIPTION
**Pull request by leoromanovsky**
…torage (FF-1980)

## Motivation and Context
[//]: #  (Why is this change required? What problem does it solve?)

* Prepare for user provided persistence stores
* 🐞 Fixes bug where stale keys in `localStorage` were not being removed. All configuration is stored to a single key.

## Description
[//]: # (Describe your changes in detail)

* bump to `commons:3.0.6`
* `LocalStorageBackedAsyncStore` implementation of `IAsyncStore` - while `localStorage` is a sync API, wrap the requests in promises to fit in with our design pattern of persistence stores all being async.
* 🐞 Fixes bug where stale keys in `localStorage` were not being removed. All configuration is stored to a single key.
* `configurationStorageFactory` detects what environment it is executing in and creates the appropriate configuration store.

## How has this been tested?
[//]: # (Please describe in detail how you tested your changes)

* unit and integration tests

[//]: # (OPTIONAL)
[//]: # (Add one or multiple labels: enhancement, refactoring, bugfix)

### CODE REVIEW

File: src/configuration-factory.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  HybridConfigurationStore,
+  IConfigurationStore,
+  MemoryOnlyConfigurationStore,
+  MemoryStore,
+} from '@eppo/js-client-sdk-common';
+
+import { LocalStorageBackedAsyncStore } from './local-storage';
+
+export function configurationStorageFactory(): IConfigurationStore<Flag> {
+  if (hasWindowLocalStorage()) {
```
Comment (leoromanovsky) [inline src/configuration-factory.ts:12]: * chrome storage will be added here
* user provided persistence store added here too

testing is a bit tricky here owing to the `window` environment. got any ideas?

File: src/local-storage-assignment-cache.spec.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
+/**
+ * @jest-environment jsdom
+ */
+
+import { LocalStorageAssignmentCache } from './local-storage-assignment-cache';
```
Comment (leoromanovsky) [inline src/local-storage-assignment-cache.spec.ts:5]: this was in `index.spec.ts` but I split it out into its own file

File: src/index.spec.ts [hunk 3]
```diff
@@ -290,283 +291,273 @@ describe('EppoJSClient E2E test', () => {
...
-      ).toEqual(true); // this key has been logged
-
-      // change variation
-      cache.setLastLoggedAssignment({
-        subjectKey: 'subject-1',
-        flagKey: 'flag-1',
-        allocationKey: 'allocation-1',
-        variationKey: 'variant',
-      });
+describe('initialization options', () => {
```
Comment (leoromanovsky) [inline src/index.spec.ts:296]: diff is a bit hard to follow in this PR but the overview is:

* moved this block into a top level function so that it is not impacted by other tests; in particular the execution of `init` in a `beforeEach` block breaks these, since they are counting network requests
* isolated all the tests

File: src/index.spec.ts [hunk 4]
```diff
@@ -290,283 +291,273 @@ describe('EppoJSClient E2E test', () => {
...
+    },
+  } as unknown as Record<'flags', Record<string, Flag>>;
+
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  let init: Function;
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  let getInstance: Function;
+
+  beforeEach(async () => {
+    jest.isolateModules(() => {
```
Comment (leoromanovsky) [inline src/index.spec.ts:313]: each test has this isolation

File: src/index.spec.ts [hunk 5]
```diff
@@ -290,283 +291,273 @@ describe('EppoJSClient E2E test', () => {
...
-      // Advance timers mid-init to allow retrying
-      await jest.advanceTimersByTimeAsync(maxRetryDelay);
+  it('default options', async () => {
+    let callCount = 0;
 
-      // Await so it can finish its initialization before this test proceeds
-      const client = await initPromise;
-      expect(callCount).toBe(2);
-      expect(client.getStringAssignment(flagKey, 'subject', {}, 'default-value')).toBe('control');
+    global.fetch = jest.fn(() => {
```
Comment (leoromanovsky) [inline src/index.spec.ts:371]: converted all `td.replace(HttpClient` to `global.fetch` mocks

File: src/local-storage.ts [hunk 6]
```diff
@@ -1,48 +1,38 @@
...
+    return this._isInitialized;
   }
 
-  public getKeys(): string[] {
-    if (hasWindowLocalStorage()) {
-      return Object.keys(window.localStorage);
-    }
-    return null;
+  isExpired(): Promise<boolean> {
+    return Promise.resolve(true);
```
Comment (schmit) [inline src/local-storage.ts:22]: should this always resolve to true?

**Review by schmit — APPROVED** [review]
Looks good to me but would love for @aarsilv to have another look as well

File: package.json [hunk 7]
```diff
@@ -57,7 +57,7 @@
     "webpack-cli": "^4.10.0"
   },
   "dependencies": {
-    "@eppo/js-client-sdk-common": "3.0.2",
+    "@eppo/js-client-sdk-common": "3.0.6",
```
[the remark this answers is not in the document]
Comment (aarsilv) [reply package.json:60]: 📈

File: src/configuration-factory.ts (hunk 1 above)
Comment (aarsilv) [reply src/configuration-factory.ts:12]: Seems like folks just mock it ([StackOverflow link](https://stackoverflow.com/a/14381941/26197))

File: src/index.spec.ts [hunk 8]
```diff
@@ -127,7 +128,7 @@ describe('EppoJSClient E2E test', () => {
   });
 
   afterEach(() => {
-    returnUfc = readMockUfcResponse;
+    //returnUfc = readMockUfcResponse;
```
Comment (aarsilv) [inline src/index.spec.ts:131]: should this just be axed then? 🪓

File: src/index.spec.ts (hunk 3 above)
Comment (aarsilv) [reply src/index.spec.ts:296]: 🙌 Heros work

File: src/index.spec.ts (hunk 4 above)
Comment (aarsilv) [reply src/index.spec.ts:313]: nice!

File: src/index.spec.ts (hunk 5 above)
Comment (aarsilv) [reply src/index.spec.ts:371]: 💪

File: src/local-storage-assignment-cache.spec.ts (hunk 2 above)
Comment (aarsilv) [reply src/local-storage-assignment-cache.spec.ts:5]: nice thinking 🧹

File: src/local-storage.ts [hunk 9]
```diff
@@ -1,48 +1,38 @@
...
-    return hasWindowLocalStorage();
-  }
+/*
+ * Removing stale keys is faciliating by storing the entire configuration into a single
+ * key in local storage. This is done by serializing the entire configuration object
+ * into a string and then storing it to a single key. When retrieving the configuration,
+ * we first check if the configuration exists in local storage. If it does, we deserialize
+ * the configuration from the string and return it. If the configuration does not exist,
+ * we return null.
+ */
```
Comment (aarsilv) [inline src/local-storage.ts:10]: ❤️ Appreciate the comment!

File: src/local-storage.ts [hunk 10]
```diff
@@ -1,48 +1,38 @@
...
+export class LocalStorageBackedAsyncStore<T> implements IAsyncStore<T> {
+  private localStorageKey = 'eppo-configuration';
+  private _isInitialized = false;
```
Comment (aarsilv) [inline src/local-storage.ts:13]: why the underscore name if its already marked as `private`?

**Review by aarsilv — APPROVED** [review]
Good stuff! Love the clean, well-tested code. 💪 

Thanks for taking the time to clean up those tests; will make subsequent development easier 🙏

**Review by ploomiz — APPROVED** [review]
